### PR TITLE
Add random tournament digraph generator.

### PR DESCRIPTION
**Pull request by shizejin**
Hi, I am working on a tournament game generator, and find that random [tournament graph](https://en.wikipedia.org/wiki/Tournament_(graph_theory)) generator has not been implemented yet in `LightGraphs`. Therefore, I wrote several lines for generating it. It would be great if you think tournament graph generator is generally useful and my code could be useful.

### CODE REVIEW

**Comment by jpfairbanks** [comment]
Looks good, can you ad a test that all edges appear in only one direction, and another test that every edge appears in at least one direction?

That will encode the definition of tournament into the code.

**Comment by mschauer** [comment]
Alternatively, one could make a function that takes a graph and returns the corresponding tournament digraph by randomly orienting the edges.

**Comment by mbesancon** [comment]
To produce a tournament graph, you need a complete undirected graph I think, this could be made in two steps (building the complete graph, then the orientation), but if you provide the second function with something else than a complete graph, it doesn't yield a tournament graph

**Comment by sbromberger** [comment]
> To produce a tournament graph, you need a complete undirected graph 

This is my understanding. Therefore

> this could be made in two steps (building the complete graph, then the orientation)

seems like a big waste of resources, as you're creating two graphs to produce one.

File: src/generators/randgraphs.jl [hunk 1]
```diff
@@ -598,6 +598,30 @@ function random_regular_digraph(n::Integer, k::Integer; dir::Symbol=:out, seed::
...
+function random_tournament_digraph(n::Integer; seed::Int=-1)
+
+    rng = getRNG(seed)
+    g = DiGraph(n)
+
+    for i = 1:n
+        for j = i:n
+            rand(rng) > 0.5 ? add_edge!(g, Edge(i, j)) : add_edge!(g, Edge(j, i))
+        end
+    end
```
Comment (sbromberger) [inline src/generators/randgraphs.jl]: In addition to adding tests, could you simplify this?
```
for i = 1:n, j = i:n
  rand(rng) > 0.5 ? add_edge!(g, Edge(i, j)) : add_edge!(g, Edge(j, i))
end
```

Comment (shizejin) [reply src/generators/randgraphs.jl]: Sure, I'll be willing to.

Comment (mschauer) [reply src/generators/randgraphs.jl]: Should also use `rand(Bool)`

Comment (sbromberger) [reply src/generators/randgraphs.jl]: Oh, yeah, good catch. @shizejin, when you're testing a coin flip, it's better to use a boolean. That way the mechanics are taken care of for you:

```
rand(Bool) ? do_if_true() : do_if_false()
```

**Comment by mschauer** [comment]
Not tournaments, but incomplete tournaments. At least I have run into those before... though I do not know if ever somebody wanted to sample them.

**Comment by shizejin** [comment]
@jpfairbanks I add [code](https://github.com/JuliaGraphs/LightGraphs.jl/pull/811/commits/d33198a6fd20c2ab24bf08a11a3c40c6ee32afc9#diff-56c866b9103c82f67e6ed61183078cfbR203) to test that the reverse of each edge does not exist in the graph. As I think `is_directed` tests that every edge has directions, this would be sufficient to ensure that each edge has exact one direction.

I modify the [code](https://github.com/JuliaGraphs/LightGraphs.jl/pull/811/commits/d33198a6fd20c2ab24bf08a11a3c40c6ee32afc9#diff-c5075146427d39c12620e0a4c5e973a5R617) so that each vertex is not directed to itself, which I think should be more natural.

**Comment by sbromberger** [comment]
> As I think `is_directed` tests that every edge has directions

It does not. `is_directed(g)` just checks the *type* of `g` - if it's a `DiGraph`, then it returns `true`. It does not look at any edges.

File: src/generators/randgraphs.jl (hunk 1 above)
Comment (shizejin) [reply src/generators/randgraphs.jl]: Thanks! It's great to learn about this. Will modify this in minutes.

File: src/generators/randgraphs.jl [hunk 2]
```diff
@@ -598,6 +598,28 @@ function random_regular_digraph(n::Integer, k::Integer; dir::Symbol=:out, seed::
     end
 end
 
+@doc_str """
+    random_tournament_digraph(n)
+
+Create a random directed [tournament graph]
+(https://en.wikipedia.org/wiki/Tournament_(graph_theory))
```
Comment (mschauer) [inline src/generators/randgraphs.jl]: remove line break to make valid markdown

File: src/generators/randgraphs.jl [hunk 3]
```diff
@@ -598,6 +598,28 @@ function random_regular_digraph(n::Integer, k::Integer; dir::Symbol=:out, seed::
...
+### Optional Arguments
+- `seed=-1`: set the RNG seed.
+"""
+function random_tournament_digraph(n::Integer; seed::Int=-1)
+
+    rng = getRNG(seed)
+    g = DiGraph(n)
+
+    for i = 1:n, j = i+1:n
+        rand(rng, Bool) ? add_edge!(g, Edge(i, j)) : add_edge!(g, Edge(j, i))
```
Comment (mschauer) [inline src/generators/randgraphs.jl]: I like more
```
add_edge!(g, rand(rng, Bool) ?  Edge(i, j) : Edge(j, i))
```

Comment (sbromberger) [reply src/generators/randgraphs.jl]: I don't like that syntax as much, since I have to go into the `add_edge` to understand there's a conditional. Is there a performance reason to do this, or is it just syntactical preference?

Comment (mschauer) [reply src/generators/randgraphs.jl]: I do not think so, but possibly `add_edge!(g, ifelse(rand(rng, Bool), Edge(i, j), Edge(j, i)))` can avoid a branch. "In some cases, using `ifelse` instead of an if statement can eliminate
  the branch in generated code and provide higher performance in tight loops." A matter of trying.

Comment (sbromberger) [reply src/generators/randgraphs.jl]: I'm not sure you can avoid branches here given that we're determining a branch based on a random number.

Comment (sbromberger) [reply src/generators/randgraphs.jl]: Let's leave as-is for now and if you're interested in experimenting we can entertain a PR modification later.

Comment (mschauer) [reply src/generators/randgraphs.jl]: Who knows, it is a matter of trying. In principle, it could be translated to branchless code (think of `i + rand(Bool)*(j-i)`).

File: src/generators/randgraphs.jl (hunk 2 above)
Comment (shizejin) [reply src/generators/randgraphs.jl]: This line break will be automatically ignored when `Documenter.jl` is generating documentations. I put line break here because otherwise it makes too long a line.

**Comment by sbromberger** [comment]
I think this looks ready. @jpfairbanks - I think the added test addresses your suggestion, but can you confirm?

File: src/generators/randgraphs.jl (hunk 2 above)
Comment (mschauer) [reply src/generators/randgraphs.jl]: Finally, escape the url:  `[tournament graph](https://en.wikipedia.org/wiki/Tournament_%28graph_theory%29)` to make the link actually working (otherwise the closing parentheses are parsed wrongly)

**Comment by jpfairbanks** [comment]
LGTM 🚢 .

**Comment by shizejin** [comment]
Thanks for reviewing!

@jpfairbanks @sbromberger I modified the [code](https://github.com/JuliaGraphs/LightGraphs.jl/pull/811/commits/c1e0292c4ec298c55b780ee40551297d38fa6c5c#diff-56c866b9103c82f67e6ed61183078cfbR203) for testing each edge has one and only one direction. If you think it looks fine, then it's ready to merge.

File: src/generators/randgraphs.jl (hunk 2 above)
Comment (mschauer) [reply src/generators/randgraphs.jl]: Ping

Comment (sbromberger) [reply src/generators/randgraphs.jl]: Yup, this needs to be done.

Comment (shizejin) [reply src/generators/randgraphs.jl]: Thanks! Modified.

**Comment by sbromberger** [comment]
Thanks, @shizejin !

**Comment by shizejin** [comment]
Thanks for all the helpful reviews and kindly suggestions! @sbromberger @jpfairbanks @mschauer
